runtime/shared_heap.c

```diff
@@ -785,6 +811,475 @@ void caml_verify_heap(caml_domain_state *domain) {
+            memcpy(new_p, p, Whsize_hd(hd) * sizeof(value));
+
+            /* Set first field of p to as a forwarding pointer */
+            Field(Val_hp(p), 0) = Val_hp(new_p);
+          }
+          /* We are implicitly sweeping pools in the evacuation set and thus
+             we must remember to call finalisers for Custom blocks. */
+          else if( Has_status_hd(hd, caml_global_heap_state.GARBAGE) )
+          {
+            /* Deal with custom block finalisers */
+            if (Tag_hd (hd) == Custom_tag) {
+              void (*final_fun)(value) = Custom_ops_val(Val_hp(p))->finalize;
+              if (final_fun != NULL) final_fun(Val_hp(p));
+            }
+          }
+        } else {
+          /* In the DEBUG runtime, we should overwrite the fields of empty
+              blocks from the 2nd field onwards (the 1st field is used for
+              the free list). */
+          #ifdef DEBUG
+          for( int w = 1 ; w < wh-1 ; w++ ) {
+            Field(Val_hp(p), w) = Debug_free_major;
+          }
+          #endif
+        }
+
+        p += wh;
+      }
+
+      evac_pool = evac_pool->next;
+    }
+
+    /* A pool we can allocate in to must have empty space we can allocate
+      a new block in to else it would should have been moved to the full pools
+      list. */
+    CAMLassert(cur_pool->next_obj != NULL);
+  }
+
+  caml_global_barrier();
+
+  /* Second phase: at this point all live blocks in evacuated pools have been
+    moved and their old location's first field now points to their new location.
+    We now go through all pools again (including full ones this time)
+    and for each field we check what shared pool it is in and if it an
+    evacuating pool. This hopefully generates fewer cache misses than using
+    a bit in the header of moved blocks. If the pool is evacuating we find
+    the moved block's forwarding pointer and update accordingly. */
+
+  /* First we do roots (locals and finalisers) */
+  caml_do_roots(&compact_update_field, 0, NULL, Caml_state, 1);
+
+  /* Next, one domain does the global roots */
+  if( participants[0] == Caml_state ) {
+    caml_scan_global_roots(&compact_update_field, NULL);
+  }
+
+  /* Shared heap pools. */
+  for(int sz_class = 1; sz_class < NUM_SIZECLASSES; sz_class++) {
+    pool* cur_pool = heap->unswept_avail_pools[sz_class];
+
+    /* Iterate through pools updating the fields in each live block,
+      stop once we reach our first evacuating pool. */
+    while( cur_pool != NULL && cur_pool->evacuating == 0 ) {
+      value* p = (value*)((char*)cur_pool + POOL_HEADER_SZ);
+      value* end = (value*)cur_pool + POOL_WSIZE;
+      mlsize_t wh = wsize_sizeclass[sz_class];
+
+      while (p + wh <= end) {
+        if( *p != 0
+          && Has_status_val(Val_hp(p), caml_global_heap_state.UNMARKED) ) {
+          compact_update_block(p);
+        }
+          p += wh;
+      }
+
+      cur_pool = cur_pool->next;
+    }
+
+    /* do the same for full pools */
```

Comment:
Can we share this code instead of repeating it?

Reply:
Fwiw the only reason it wasn't shared was because we did an extra assert for full pools.

We've removed that assert and shared the code now.